Symbolic names for ImageBuffer backends complicate the code unneccessarily

kkinnunen-apple opened the pull request:
#### 18cbc28a15bebf4e20edcfadd97f145b85d008f8
<pre>
Symbolic names for ImageBuffer backends complicate the code unneccessarily
<a href="https://bugs.webkit.org/show_bug.cgi?id=262110">https://bugs.webkit.org/show_bug.cgi?id=262110</a>
rdar://116049610

Reviewed by Said Abou-Hallawa.

The symbolic names UnacceleratedImageBuffer*Backend and
AcceleratedImageBuffer*Backend were probably originally introduced
to help the caller code to create image buffers in cross-platform way.
However, the image buffer instantiation is based on global policy that
is invoked by ImageBuffer::create().

Instantiating the Accelerated image buffers is platform-specific operation,
as demonstrated by IOSurface specific parameters in ImageBufferCreationContext
amalgamation structure.

Remove the symbolic names, as they make further simplification unneccessarily
complex. Just use the normal class names inside ifdefs that already exist
at the call sites.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::create):
* Source/WebCore/platform/graphics/PlatformImageBuffer.h:
* Source/WebCore/platform/graphics/PlatformImageBufferBackend.h: Removed.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::createImageBuffer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
(WebKit::RemoteLayerBackingStoreCollection::allocateBufferForBackingStore):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.cpp:
(WebKit::ImageBufferShareableBitmapBackend::ImageBufferShareableBitmapBackend):
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/PlatformImageBufferShareableBackend.h: Removed.
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::didCreateBackend):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::createImageBuffer):
* Source/WebKit/WebProcess/WebPage/wc/DrawingAreaWC.cpp:
(WebKit::DrawingAreaWC::createImageBuffer):

Canonical link: <a href="https://commits.webkit.org/268563@main">https://commits.webkit.org/268563@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b5ddfd4169edb951d61fd0d11ee456cd2d45df30

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19905 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20329 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20949 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21798 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18591 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20141 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23586 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20478 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20115 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20127 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20097 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17311 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22652 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17268 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18106 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24374 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18344 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18282 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22360 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18876 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15999 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18047 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4805 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22396 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18700 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/PlatformImageBuffer.h:
```diff
@@ -40,15 +40,9 @@
 
 #if USE(CG)
-using UnacceleratedImageBufferBackend = ImageBufferCGBitmapBackend;
+using ImageBufferPlatformBitmapBackend = ImageBufferCGBitmapBackend;
```
If this file contains a single definition which is `ImageBufferPlatformBitmapBackend`, then I think this file should be renamed `ImageBufferPlatformBitmapBackend.h`

shallawa commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
```diff
@@ -59,6 +59,11 @@
+#if HAVE(IOSURFACE)
+#include "ImageBufferRemoteIOSurfaceBackend.h"
+#include "ImageBufferShareableMappedIOSurfaceBackend.h"
+#endif
```
Are these backends used in this source file?

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.h:
```diff
@@ -38,7 +43,13 @@ namespace WebKit {
 class ShareableBitmap;
 
-class ImageBufferShareableBitmapBackend final : public WebCore::PlatformImageBufferBackend, public ImageBufferBackendHandleSharing {
+#if USE(CG)
+using ImageBufferShareableBitmapBackendBase = WebCore::ImageBufferCGBackend;
+#elif USE(CAIRO)
+using ImageBufferShareableBitmapBackendBase = WebCore::ImageBufferCairoBackend;
+#endif
```
This is a little bit awkward. 

1. Both `ImageBufferCGBackend` and `ImageBufferCairoBackend` are WebCore objects. So why do we need to have them redefined as `ImageBufferShareableBitmapBackendBase` in WebKit?
2. Both do not handle any backend sharing.
3. `ImageBufferCGBackend` is not necessarily unaccelerated. It the base of `ImageBufferIOSurfaceBackend` also.

Can't we keep them in WebCore under the name ImageBufferPlatformBackend?

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
```diff
@@ -205,13 +210,17 @@ void RemoteImageBufferProxy::didCreateBackend(std::optional<ImageBufferBackendHa
+        backend = ImageBufferShareableBitmapBackend::create(parameters(), WTFMove(*handle));
+#else
+    backend = ImageBufferShareableBitmapBackend::create(parameters(), WTFMove(*handle));
+#endif
```
This code can be shortened 

```
    if (renderingMode() == RenderingMode::Accelerated) {
#if HAVE(IOSURFACE)
        if (canMapBackingStore())
            backend = ImageBufferShareableMappedIOSurfaceBackend::create(parameters(), WTFMove(*handle));
        else
            backend = ImageBufferRemoteIOSurfaceBackend::create(parameters(), WTFMove(*handle));
#endif 
    } else
        backend = ImageBufferShareableBitmapBackend::create(parameters(), WTFMove(*handle));
```

Cairo ports should not be setting the renderingMode to RenderingMode::Accelerated.

kkinnunen-apple committed:
Symbolic names for ImageBuffer backends complicate the code unneccessarily
https://bugs.webkit.org/show_bug.cgi?id=262110
rdar://116049610

Reviewed by Said Abou-Hallawa.

The symbolic names UnacceleratedImageBuffer*Backend and
AcceleratedImageBuffer*Backend were probably originally introduced
to help the caller code to create image buffers in cross-platform way.
However, the image buffer instantiation is based on global policy that
is invoked by ImageBuffer::create().

Instantiating the Accelerated image buffers is platform-specific operation,
as demonstrated by IOSurface specific parameters in ImageBufferCreationContext
amalgamation structure.

Remove the symbolic names, as they make further simplification unneccessarily
complex. Just use the normal class names inside ifdefs that already exist
at the call sites.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::create):
* Source/WebCore/platform/graphics/PlatformImageBuffer.h:
* Source/WebCore/platform/graphics/PlatformImageBufferBackend.h: Removed.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::createImageBuffer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
(WebKit::RemoteLayerBackingStoreCollection::allocateBufferForBackingStore):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.cpp:
(WebKit::ImageBufferShareableBitmapBackend::ImageBufferShareableBitmapBackend):
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/PlatformImageBufferShareableBackend.h: Removed.
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::didCreateBackend):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::createImageBuffer):
* Source/WebKit/WebProcess/WebPage/wc/DrawingAreaWC.cpp:
(WebKit::DrawingAreaWC::createImageBuffer):

Canonical link: https://commits.webkit.org/268563@main
